### PR TITLE
[dhcpmon] Fix dhcpmon socket filter and tx count issue

### PR DESCRIPTION
**Pull request by kellyyeh**
#### Why I did it
1. Fix issue caused by dualtor support PR https://github.com/sonic-net/sonic-buildimage/pull/11201
2. Improve code

#### How I did it
1. On single ToR, packets received count was duplicated due to socket filter set to "inbound"
2. Tx count not increasing due to filter set to "inbound". Added an outbound socket to count tx packets
3. Added vlan member interface mapping for Ethernet interface to vlan interface lookup in reference to PR https://github.com/sonic-net/sonic-dhcp-relay/pull/27
4. Exit when socket fails to initialize to allow dhcp_relay docker to restart

#### How to verify it
Tested on vstestbed single tor and dual tor, sent packets and verify printed out dhcpmon rx and tx counters is correct
- Correct number of tx increases
- Tx does not increase when ToR is on standby

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012
- [ ] 202106
- [ ] 202111
- [x] 202205
- [ ] 202211

### CODE REVIEW

File: src/dhcpmon/src/dhcp_device.cpp [hunk 1]
```diff
@@ -163,86 +213,102 @@ static void handle_dhcp_option_53(dhcp_device_context_t *context,
...
+                                    ethhdr->ether_shost[3] == context->mac[3] &&
+                                    ethhdr->ether_shost[4] == context->mac[4] &&
+                                    ethhdr->ether_shost[5] == context->mac[5]) ?
+                                    DHCP_TX : DHCP_RX;
+        int offset = 0;
+        int stop_dhcp_processing = 0;
+        while ((offset < (dhcp_option_sz + 1)) && dhcp_option[offset] != 255) {
+            switch (dhcp_option[offset])
+            {
+            case 53:
```
Comment (jcaiMR) [inline src/dhcpmon/src/dhcp_device.cpp:252]: Better to use enum for option53, like:
enum {
  OptionDHCPMessageType = 53,
}
If we only care about option53, maybe we don;t need switch case here, just use if and remove temp variable stop_dhcp_processing.

**Comment by jcaiMR** [comment]
I think current diff still has duplicate packets issue, because listen socket bind on all interfaces.
So each time call dhcp_devman_add_intf( ) we will have a bind all input socket, and input packets will reach each all sockets.

Normally, for bind all interface, we just need one socket and give this socket big enough rcvd buffer size.
For TX socket as it bind to particular interface, so each interface a socket would be fine. 

And if we use one socket for all input interface binding, better to use batch process for better performance, that is one fd will trigger a batch recvfrom.

**Comment by kellyyeh** [comment]
Each dhcpmon process will only monitor one vlan
https://github.com/sonic-net/sonic-buildimage/blob/3740f1efb7be8be325e93f37834adf7cc61e169f/dockers/docker-dhcp-relay/dhcp-relay.monitors.j2#L44

File: src/dhcpmon/src/dhcp_device.cpp [hunk 2]
```diff
@@ -252,62 +308,28 @@ static void read_callback_dual_tor(int fd, short event, void *arg)
     while ((event == EV_READ) &&
```
Comment (jcaiMR) [inline src/dhcpmon/src/dhcp_device.cpp:308]: Here "event == EV_READ" will always true as we register this handler for EV_READ, we don't need to check it again.

File: src/dhcpmon/src/dhcp_device.cpp [hunk 3]
```diff
@@ -163,86 +213,92 @@ static void handle_dhcp_option_53(dhcp_device_context_t *context,
...
  *
  * @return none
  */
-static void read_callback(int fd, short event, void *arg)
+static void read_tx_callback(int fd, short event, void *arg)
 {
     dhcp_device_context_t *context = (dhcp_device_context_t*) arg;
     ssize_t buffer_sz;
 
     while ((event == EV_READ) &&
```
Comment (jcaiMR) [inline src/dhcpmon/src/dhcp_device.cpp:284]: Here "event == EV_READ" will always true as we register this handler for EV_READ, we don't need to check it again.

File: src/dhcpmon/src/dhcp_device.cpp [hunk 4]
```diff
@@ -495,18 +517,29 @@ static int init_socket(dhcp_device_context_t *context, const char *intf)
...
+        }
+
+        struct sockaddr_ll tx_addr;
+        memset(&tx_addr, 0, sizeof(tx_addr));
+        tx_addr.sll_ifindex = if_nametoindex(intf);
+        tx_addr.sll_family = AF_PACKET;
+        tx_addr.sll_protocol = htons(ETH_P_ALL);
+        if (bind(context->tx_sock, (struct sockaddr *) &tx_addr, sizeof(tx_addr))) {
             syslog(LOG_ALERT, "bind: failed to bind to interface '%s' with '%s'\n", intf, strerror(errno));
             break;
```
Comment (jcaiMR) [inline src/dhcpmon/src/dhcp_device.cpp:544]: If we failed to bind socket, we may just call exit() to let container restart.

File: src/dhcpmon/src/dhcp_device.cpp [hunk 5]
```diff
@@ -495,18 +517,29 @@ static int init_socket(dhcp_device_context_t *context, const char *intf)
     int rv = -1;
 
     do {
-        context->sock = socket(AF_PACKET, SOCK_RAW | SOCK_NONBLOCK, htons(ETH_P_ALL));
-        if (context->sock < 0) {
+        context->rx_sock = socket(AF_PACKET, SOCK_RAW | SOCK_NONBLOCK, htons(ETH_P_ALL));
+        context->tx_sock = socket(AF_PACKET, SOCK_RAW | SOCK_NONBLOCK, htons(ETH_P_ALL));
+        if (context->rx_sock < 0 || context->tx_sock < 0) {
             syslog(LOG_ALERT, "socket: failed to open socket with '%s'\n", strerror(errno));
             break;
```
Comment (jcaiMR) [inline src/dhcpmon/src/dhcp_device.cpp:524]: if socket failed to create, we can just exit program.

File: src/dhcpmon/src/dhcp_device.cpp [hunk 6]
```diff
@@ -666,21 +700,27 @@ int dhcp_device_start_capture(dhcp_device_context_t *context,
         }
         context->snaplen = snaplen;
 
-        if (setsockopt(context->sock, SOL_SOCKET, SO_ATTACH_FILTER, &dhcp_sock_bfp, sizeof(dhcp_sock_bfp)) != 0) {
+        if (setsockopt(context->rx_sock, SOL_SOCKET, SO_ATTACH_FILTER, &dhcp_inbound_sock_bfp, sizeof(dhcp_inbound_sock_bfp)) != 0) {
             syslog(LOG_ALERT, "setsockopt: failed to attach filter with '%s'\n", strerror(errno));
```
Comment (jcaiMR) [inline src/dhcpmon/src/dhcp_device.cpp:704]: Same like above, maybe we just exit program if any error happened in this function.

File: src/dhcpmon/src/dhcp_device.cpp (hunk 3 above)
Comment (kellyyeh) [reply src/dhcpmon/src/dhcp_device.cpp:284]: event could be EV_PERIST or EV_READ

File: src/dhcpmon/src/dhcp_device.cpp (hunk 4 above)
Comment (kellyyeh) [reply src/dhcpmon/src/dhcp_device.cpp]: Updated

File: src/dhcpmon/src/dhcp_device.cpp (hunk 5 above)
Comment (kellyyeh) [reply src/dhcpmon/src/dhcp_device.cpp]: Updated

File: src/dhcpmon/src/dhcp_device.cpp (hunk 1 above)
Comment (kellyyeh) [reply src/dhcpmon/src/dhcp_device.cpp]: Updated

File: src/dhcpmon/src/dhcp_device.cpp (hunk 3 above)
Comment (jcaiMR) [reply src/dhcpmon/src/dhcp_device.cpp:284]: EV_PERIST controls the event is persistent, it makes you don't need to use event_add() again after callback. But when callback is called it must be a READ event, as you don't register read_tx_callback to any WRITE/DELETE events.

Comment (kellyyeh) [reply src/dhcpmon/src/dhcp_device.cpp]: Thanks I see. Removed event check

**Review by jcaiMR — APPROVED** [review]

**Review by yxieca — APPROVED** [review]

**Comment by qiluo-msft** [comment]
What issue is fixed by this PR? could you explain in PR description?

**Comment by yxieca** [comment]
@kellyyeh  is there a particular reason you didn't request to backport to 202211 branch?

**Comment by yxieca** [comment]
@kellyyeh please create PR for 202205 branch for identify missing dependencies

**Comment by kellyyeh** [comment]
@yxieca PR https://github.com/sonic-net/sonic-buildimage/pull/11201 is not cherry-picked to 202211 yet. Just added label for both PRs

**Comment by qiluo-msft** [comment]
This commit could not be cleanly cherry-picked to 202012. Please submit another PR.

**Comment by yxieca** [comment]
@kellyyeh this change cannot be cherry-picked cleanly to 20205 branch. can you raise separate PR?

**Comment by liuh-80** [comment]
202012 branch PR merged: https://github.com/sonic-net/sonic-buildimage/pull/13441

**Pull request merged**